metals/src/main/scala/scala/meta/internal/metals/ProcessOutput.scala

```diff
@@ -0,0 +1,74 @@
+package scala.meta.internal.metals
+
+import java.lang.StringBuilder
+import java.nio.ByteBuffer
+import java.nio.charset.StandardCharsets
+import fansi.ErrorMode
+
+/**
+ * Handles streaming console output from a system process with potential ANSI codes.
+ *
+ * @param onLine The callback handler when the process has printed a single line.
+ *               Guaranteed to have no newline \n characters.
+ */
+class ProcessOutput(onLine: String => Unit) {
+  private var buffer = new StringBuilder()
+
+  /** The process has exited, clears out buffered output. */
+  def onProcessExit(): Unit = {
+    if (buffer.length() > 0) {
+      onLine(buffer.toString())
+      reset()
+    }
+  }
+
+  def onByteOutput(bytes: ByteBuffer): this.type = {
+    onPlainOutput(toPlainString(bytes))
+  }
+
+  def onStringOutput(text: String): this.type = {
+    onPlainOutput(toPlainString(text))
+  }
+
+  private def onPlainOutput(text: String): this.type = {
+    def loop(start: Int): Unit = {
+      val newline = text.indexOf('\n', start)
+      if (newline < 0) {
+        buffer.append(text, start, text.length())
+      } else {
+        onLine(lineSubstring(text, start, newline))
+        loop(newline + 1)
+      }
+    }
+    loop(0)
+    this
+  }
+
+  private def lineSubstring(text: String, from: Int, to: Int): String = {
```

Comment:
Instead we get  two methods and a recursion which is a bit more complex to reason about.

Reply:
The recursive implementation may not be as readable but it achieves several performance improvements

- less copying, we perform the minimal number of `substring` operations (one per line)
- fewer redundant operations, the recursive loop iterates over each character at most once

This is fairly performance sensitive code since we're feeding the standard output of a build process (downloading dependencies, ...) to the LSP client.